modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ExchangeContext.java

```diff
@@ -58,7 +66,17 @@
     public ExchangeContext(boolean crd, GridDhtPartitionsExchangeFuture fut) {
         int protocolVer = exchangeProtocolVersion(fut.firstEventCache().minimumNodeVersion());
 
-        if (compatibilityNode || (crd && fut.localJoinExchange())) {
+        GridDiscoveryManager disco = fut.sharedContext().discovery();
+
+        if (protocolVer > 2 && fut.exchangeId().isLeft() && fut.isEventNodeInBaseline() &&
```

Comment:
This protocol works only if cluster is in-memory and baseline-autoadjust is enabled.
How it works when persistence is enabled and with/without baseline autoadjust is enabled?

Reply:
What protocol do you mean? 
protocolVer == 3 was introduced by me.

Reply:
Yes, I mean the protocol introduced by yourself.

Reply:
`disco.baselineChanged(fut.sharedContext().exchange().readyAffinityVersion(), fut.initialVersion())`
This code will return `true` only if baseline has changed by `NODE_LEFT` event. Baseline is changed automatically in discovery cache only if it's in-memory cluster and auto-adjust is enabled. In other cases baseline is changed manually (if auto-adjust disabled) or by separate event (ChangeGlobalState) right after topology event if auto-adjust is enabled.

Reply:
It still looks like code works, but disco.baselineChanged(...) can be used on in case of nodeFail, correct?
Persistence were checked by TxPartitionCounterStateConsistencyTest#testPartitionConsistencyWithPrimaryRestart. 
As you told, persistence case (auto-adjust in [true, false]) on nodeLeft always leads to recovery instead of regular PME.

I don't like the idea to have a check like persistence.enabled() here, because we should check the result (was baseline changed or not) instead of checking reasons leads to this (which, definitely, will be changed in future, and can be (already) overcomplicated).
Could you help with writing a simple and proper "if" statement?

Reply:
I think, there should be no relying on the changed baseline or not.
Why do you check it here?

Reply:
Because "recovery only PME" possible only in case affinity was not changed, and only primary-backup switches are expected.

Reply:
The primary-backup switch is only possible when baseline is not changed :)

Reply:
That's exactly the case covered by this fix :)

Reply:
My bad :) 

public boolean baselineChanged(AffinityTopologyVersion topVer1, AffinityTopologyVersion topVer2) {
...
return Objects.equals(baseline1, baseline2); 
}

Reply:
Issue solved?